Add Mastodon package and module

happy-river opened the pull request:
###### Motivation for this change

This PR continues the work done in #60788 by @petabyteboy, and adds these improvements:

- Add a shell script to set up the environment variables for `tootctl`.
- Make database setup work correctly with Mastodon 3.0.1.
- Allow database and SMTP passwords to contain spaces.
- Improve descriptions of module options.
- Make `pkgs.mastodon.override { ... }` work to build forks of Mastodon.
- Add a `package` option to the service module.
- Borrow a fix from #76765 to make the package work with rubygems 3.1.2.
- Remove the `import <nixpkgs>` from the source expression.
- Make the update script work, and add a derivation for it.
- Add documentation.

Things that could still be improved:

- [ ] Investigate if it makes sense to allow building, enabling and disabling the three services (web, sidekiq and streaming) separately, similar to the kubernetes module.
- [ ] Add more advanced options, i.e. the number of sidekiq threads, support for S3 storage backend, ...

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
mastodon: init at 3.0.1

Commit:
nixos/mastodon: init

nixos/mastodon: add webserver test
nixos/mastodon: add package tests
Tests and fixes contributed by @happy-river, thanks!

happy-river committed:
nixos/mastodon: add mastodon-env shell script to support tootctl

happy-river committed:
nixos/mastodon: support passwords containing spaces

happy-river committed:
nixos/mastodon: seed database on initial creation

happy-river committed:
nixos/mastodon: improve descriptions of options

happy-river committed:
maintainers: add happy-river

happy-river committed:
mastodon: add override mechanism for src

happy-river committed:
nixos/mastodon: add package option

happy-river committed:
mastodon: make compatible with rubygems 3.1.2

happy-river committed:
mastodon: remove use of <nixpkgs> and fix update script

happy-river committed:
nixos/mastodon: make it easier to run 'rake db:migrate'

happy-river committed:
nixos/mastodon: add documentation

volth commented on pkgs/servers/mastodon/source.nix:
```diff
@@ -0,0 +1,14 @@
+    sha256 = "11vljzarhgs6hik6qhimlaimqm5ysljf9jm402r5fpy6axh6yksc";
+  };
+  version = import ./version.nix;
+in runCommand "mastodon-source" { buildInputs = [ jq ]; } ''
```
btw there is `applyPatches` function for that

volth replied:
or even `postFetch` parameter of `fetchgit`

happy-river replied:
Thanks for the suggestions. I reworked it to use `applyPatches`.

Mic92 commented:
cc @schmittlauch

Mic92 commented:
Hey. This looks good to me. Is it feasible though to add mastodons npm packages to
our node-packages set though? We are already doing this for other packages like lumo or iosevka: https://github.com/NixOS/nixpkgs/blob/8539d5f48f9de4ba428d2ca6687712db1d9cb2b7/pkgs/development/node-packages/node-packages-v10.json#L80

petabyteboy commented:
> Hey. This looks good to me. Is it feasible though to add mastodons npm packages to
> our node-packages set though? We are already doing this for other packages like lumo or iosevka:
> 
> https://github.com/NixOS/nixpkgs/blob/8539d5f48f9de4ba428d2ca6687712db1d9cb2b7/pkgs/development/node-packages/node-packages-v10.json#L80

Afaik there are no official user-targeted mastodon npm packages. The webapp itself is not an npm package either, but it is exposed as a nix derivation in the nixpkgs attrset (as pkgs.mastodon.mastodon-assets).

Mic92 commented:
@petabyteboy The idea is that we can share npm package source between different packages in nixpkgs instead of having 11.000 lines of generated nix expressions per package, which affects also evaluation time.

petabyteboy commented:
> @petabyteboy The idea is that we can share npm package source between different packages in nixpkgs instead of having 11.000 lines of generated nix expressions per package, which affects also evaluation time.

Ah I see. But I doubt this would be possible since mastodon is using yarn and not npm. 
It would be nice to see support for this kind of optimization in yarn2nix, possibly also sharing the node packages expression with node2nix.

Mic92 commented:
> Ah I see. But I doubt this would be possible since mastodon is using yarn and not npm.
> It would be nice to see support for this kind of optimization in yarn2nix, possibly also sharing the node packages expression with node2nix.

Do both not use a `packages.json`. It thought that this could be read by npm as well.

petabyteboy commented:
> Do both not use a `packages.json`. It thought that this could be read by npm as well.

Yes and no, the mastodon repository contains a yarn.lock file with extra version information that can not be described in the original package.json format. NPM added a package-lock.json file to support the same thing, but it is incompatible with yarn.

Mic92 commented:
> > Do both not use a `packages.json`. It thought that this could be read by npm as well.
> 
> Yes and no, the mastodon repository contains a yarn.lock file with extra version information that can not be described in the original package.json format. NPM added a package-lock.json file to support the same thing, but it is incompatible with the yarn.lock format.

So it won't capture the some indirect dependencies pinnings? Does this matter in our case though?

petabyteboy commented:
> So it won't capture the some indirect dependencies pinnings? Does this matter in our case though?

I think so. Even if it is possible to package the dependencies with node2nix, the dependency versions could change every time the nix expression is generated. With such a complex application I think it is crucial to reproduce the dependencies as close to what upstream intends, and they clearly instruct users to use yarn when building the frontend.

Mic92 commented:
But on the other hand we also need to keep nixpkgs usuable. This packages adds 149kb of data to nixpkgs that everyone needs to download. I get regularly complaints of people not being able to run `nixpkgs-review` with local evaluation because lack of RAM.

happy-river commented:
>        expected: "sr-Latn"
>             got: "sr"

Could this be leakage from your environment into the test environment? What is your `i18n.defaultLocale`?

petabyteboy commented:
> > ```
> >    expected: "sr-Latn"
> >         got: "sr"
> > ```
> 
> Could this be leakage from your environment into the test environment? What is your `i18n.defaultLocale`?

No, this output is from the NixOS test you wrote.

happy-river commented:
> > > ```
> > >    expected: "sr-Latn"
> > >         got: "sr"
> > > ```
> > 
> > 
> > Could this be leakage from your environment into the test environment? What is your `i18n.defaultLocale`?
> 
> No, this output is from the NixOS test you wrote.

On further investigation, it appears to be a bug in the Mastodon test which happens rarely due to the use of randomization in the test. I'll report it upstream.

volth commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,537 @@
+      ] else []
+    ) env))));
+
+  mastodonEnv = pkgs.runCommand "mastodon-env" { preferLocalBuild = true; } ''
```
`pkgs.writeShellScriptBin`

happy-river replied:
Fixed, thanks for the suggestion.

nyanloutre commented:
After some testing there should be an option to disable SMTP_LOGIN and SMTP_PASSWORD as it will block access to local smtp server without authentication.
Even with auth method to none like in the following example it will not work if the login and password variables are defined:
```
SMTP_AUTH_METHOD = "none";
SMTP_OPENSSL_VERIFY_MODE = "none";
```

happy-river committed:
mastodon: use applyPatches instead of runCommand in source.nix

happy-river committed:
nixos/mastodon: use writeShellScriptBin

happy-river committed:
nixos/mastodon: fix indentation

happy-river committed:
mastodon: set supported platforms

happy-river committed:
nixos/mastodon: set random seed for package tests

happy-river commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,537 @@
+
+  envFile = pkgs.writeText "mastodon.env" (lib.concatMapStrings (s: s + "\n") (
+    (lib.concatLists (lib.mapAttrsToList (name: value:
+     if value != null then [
```
Fixed.

happy-river commented:
> After some testing I figured there should be an option to disable SMTP_LOGIN and SMTP_PASSWORD as it will block access to local smtp server without authentication.

Does this work?

```
services.mastodon.extraConfig = {
    SMTP_LOGIN = "";
    SMTP_PASSWORD = "";
};
```

If not, please share your local smtp server configuration and steps to reproduce the problem.

nyanloutre commented:
According to the documentation, these options must be commented and not just left blank

happy-river commented:
My point was that you can set those two variables to whatever you want using `services.mastodon.extraConfig`, and the values you set there will override the values set by the other module options.

ashkitten commented:
i believe the issue here is that the values in question need to not only be blank, they need to be *unset* - as in not declared at all.

happy-river committed:
nixos/mastodon: add smtp.authenticate option

happy-river commented:
In today's force push, I have:

- Removed my attempt to remove Mastodon's `package.json` from Nixpkgs. I'm not sure why it has to be in the package directory instead of using the one in `src`, but it broke one of the `grahamcofborg-eval` tests.  It was an attempt to reduce the amount of data this package adds to Nixpkgs, but it would have only saved 6kb.
- Added a new option `smtp.authenticate`, which if set to false will leave `SMTP_LOGIN` and `SMTP_PASSWORD` unset.
- Fixed some indentation and a typo in the documentation.
- Improved error reporting from `update.sh`.

I believe I've now addressed all the issues which have been brought up here except for the volume of automatically generated Ruby and JS package definitions in `yarn.nix` and `gemset.nix`. Many of the automatically generated definitions added by this package are duplicates of those in other existing packages in Nixpkgs which use `yarn2nix` and/or `bundix`, but I think the duplication problem would be best handled by a separate issue aiming to fix it for all the users of `bundix` and `yarn2nix` at once.

petabyteboy commented:
> * Removed my attempt to remove Mastodon's `package.json` from Nixpkgs. I'm not sure why it has to be in the package directory instead of using the one in `src`, but it broke one of the `grahamcofborg-eval` tests.  It was an attempt to reduce the amount of data this package adds to Nixpkgs, but it would have only saved 6kb.

The way yarn2nix works it converts the contents of package.json to a nix expression and uses those values for example to set the version attribute of the derivation. It is necessary to build the source derivation to determine the derivation attributes, making it a kind of import from derivation thing which is disabled on hydra and ofborg.

nyanloutre commented:
I tried to use the update.sh script to get a newer mastodon version, but I get this error when trying to build mastodon 3.1.2:
```
building '/nix/store/adzyx91p9s3fb16zkln18c7vipz2qnii-mastodon-assets-3.1.2.drv'...
unpacking sources
unpacking source archive /nix/store/a7gd4jaqqafqv6qz832znxwyl57r2wvp-mastodon-patched
source root is mastodon-patched
patching sources
configuring
no configure script, doing nothing
building
yarn install v1.21.1
warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.
warning Skipping preferred cache folder "/homeless-shelter/.cache/yarn" because it is not writable.
warning Selected the next writable cache folder in the list, will be "/build/.yarn-cache-1000".
info There appears to be trouble with your network connection. Retrying...
[1/6] Validating package.json...
[2/6] Resolving packages...
[3/6] Fetching packages...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
error An unexpected error occurred: "https://registry.yarnpkg.com/@babel/core/-/core-7.8.4.tgz: getaddrinfo ENOTFOUND registry.yarnpkg.com registry.yarnpkg.com:443".
info If you think this is a bug, please open a bug report with the information provided in "/build/mastodon-patched/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
builder for '/nix/store/adzyx91p9s3fb16zkln18c7vipz2qnii-mastodon-assets-3.1.2.drv' failed with exit code 1
cannot build derivation '/nix/store/46v5y9gqndyjim789kjxry1gnh3mm2pp-mastodon-3.1.2.drv': 1 dependencies couldn't be built
error: build of '/nix/store/46v5y9gqndyjim789kjxry1gnh3mm2pp-mastodon-3.1.2.drv' failed
```

happy-river commented:
> I tried to use the update.sh script to get a newer mastodon version, but I get this error when trying to build mastodon 3.1.2:
> 

I've observed the same error but haven't had time yet to investigate what is going wrong.

ashkitten commented:
it seems for forks that don't tag releases using semver (or at all) it seems we need a dummy version number or yarn will error out.

ashkitten commented:
> I tried to use the update.sh script to get a newer mastodon version, but I get this error when trying to build mastodon 3.1.2:

this seems to be due to a bump in the webpacker version used by mastodon.. possibly [this commit](https://github.com/rails/webpacker/commit/bbd0bce0dc761a34237ae831c20cc3bcfb0c0a14) in webpacker?

reading the source code of webpacker, [this seems to suggest](https://github.com/rails/webpacker/blob/master/lib/tasks/webpacker/compile.rake) that `assets:precompile` is functionally equivalent to `webpacker:compile` preceded by `webpacker:yarn_install`. maybe we should call `webpacker:compile` instead?

ashkitten commented:
paperclip seems to think every image is spoofing its mime-type because `file` is broken:

```
Mar 10 03:04:00 mastodon sidekiq[542]: Command :: file -b --mime '/tmp/8d777f385d3dfec8815d20f7496026dc20200310-542-1hamoti'
Mar 10 03:04:00 mastodon sidekiq[542]: [paperclip] Trying to link /tmp/8d777f385d3dfec8815d20f7496026dc20200310-542-1hamoti to /tmp/af89efbec357bd2465f89d42de4bc46420200310-542-127xtnf.png
Mar 10 03:04:00 mastodon sidekiq[542]: Command :: file -b --mime '/tmp/af89efbec357bd2465f89d42de4bc46420200310-542-127xtnf.png'
Mar 10 03:04:00 mastodon sidekiq[542]: [paperclip] Content Type Spoof: Filename 12aca345e22301fd.png ( from Headers, ["image/png"] from Extension), content type discovered from file command: . See documentation to allow this combination.
Mar 10 03:04:00 mastodon sidekiq[542]: [paperclip] Trying to link /tmp/8d777f385d3dfec8815d20f7496026dc20200310-542-1hamoti to /tmp/af89efbec357bd2465f89d42de4bc46420200310-542-pu5ekk.png
Mar 10 03:04:00 mastodon sidekiq[542]: Command :: file -b --mime '/tmp/af89efbec357bd2465f89d42de4bc46420200310-542-pu5ekk.png'
Mar 10 03:04:00 mastodon sidekiq[542]: [paperclip] Content Type Spoof: Filename 12aca345e22301fd.png ( from Headers, ["image/png"] from Extension), content type discovered from file command: . See documentation to allow this combination.
```

ashkitten commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,541 @@
+        PrivateTmp = true;
+        LogsDirectory = "mastodon";
+        StateDirectory = "mastodon";
+      };
```
```suggestion
      };
      path = with pkgs; [ file imagemagick ffmpeg ];
```

sidekiq needs these dependencies too

nyanloutre replied:
This fixed thumbnails and profile pictures not downloaded by the server :+1:

ashkitten commented on pkgs/servers/mastodon/default.nix:
```diff
@@ -0,0 +1,109 @@
+          cp -r ${mastodon-js-modules}/libexec/*/*/node_modules node_modules
+      fi
+      chmod -R u+w node_modules
+      rake assets:precompile
```
```suggestion
      rake webpacker:compile
```

doesn't try to run `yarn install`

happy-river replied:
Thanks for the suggestion. I had to make the same change in the test which runs Mastodon's test suite.

erictapen commented:
While this PR is still getting reviewed, I'd love to test this code with my current flake based setup. Could you rebase to current master, so the recently merged `flake.nix` gets available? That would make integration easier for me.

erictapen commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,541 @@
+        SECRET_KEY_BASE=$(cat ${cfg.secretKeyBaseFile})
+        OTP_SECRET=$(cat ${cfg.otpSecretFile})
+        VAPID_PRIVATE_KEY=$(cat ${cfg.vapidPrivateKeyFile})
+        VAPID_PUBLIC_KEY=$(cat ${cfg.vapidPublicKeyFile})
```
Every variable here must be enclosed in double quotes as it is done with `DB_PASS`. I had one of my base64 encoded secrets end in a `=` and that messed up my environment.

happy-river replied:
Fixed.

erictapen commented on nixos/modules/services/web-apps/mastodon.nix:
```diff
@@ -0,0 +1,541 @@
+
+      localDomain = lib.mkOption {
+        description = "The domain serving your Mastodon instance.";
+        default = "social.example.org";
```
This should be `example = "social.example.org";` imho. I see no reason to have an default domain here.

happy-river replied:
Fixed.

happy-river committed:
mastodon: 3.0.1 -> 3.1.2

happy-river committed:
nixos/mastodon: make dependencies available to Sidekiq

Co-Authored-By: ash lea <ashkitten@users.noreply.github.com>

happy-river committed:
nixos/mastodon: wrap base64-encoded secrets in quotes

happy-river committed:
nixos/mastodon: provide example domain instead of default domain

happy-river commented:
The latest force push updates Mastodon to version 3.1.2 and incorporates the suggestions made here in the last few weeks.

ashkitten commented:
looks good, except we still need to ensure that the node package version follows semver (in case we're passing a git hash to the update script and not semver based tags)

happy-river commented:
> looks good, except we still need to ensure that the node package version follows semver (in case we're passing a git hash to the update script and not semver based tags)

You're right, I missed that one.  I'll see what I can do.

ashkitten commented:
we don't need to have a real node package version anyways, as long as it follows semver. i don't think the node package version is used anywhere, just by mkYarnPackage.

erictapen commented:
Apparently Ruby gems can be unpublished? I can't download `sidekiq-unique-jobs-6.0.18.gem` anymore, but putting a `6.0.19` in the URL works.

```
trying https://rubygems.org/gems/sidekiq-unique-jobs-6.0.18.gem
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
^M  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0^M  0     0    0     0    0  >
curl: (22) The requested URL returned error: 403 Forbidden
error: cannot download sidekiq-unique-jobs-6.0.18.gem from any mirror
```

happy-river committed:
mastodon: use working version of sidekiq-unique-jobs gem

Use Mastodon upstream's patch to change the version of
sidekiq-unique-jobs to 6.0.20, because 6.0.18 was withdrawn.  This
commit contains manual changes to automatically generated files, and
these changes should not be needed for versions of Mastodon after
v3.1.2.

happy-river commented:
> Apparently Ruby gems can be unpublished? I can't download `sidekiq-unique-jobs-6.0.18.gem` anymore, but putting a `6.0.19` in the URL works.

Mastodon upstream has a fix for this, so I added their patch to `source.nix` and updated `gemset.nix`, by manual edits to automatically generated files, which means this change will be overwritten when `update.sh` is run for the next version of Mastodon, which is fine because it will no longer be needed.

nyanloutre commented:
I tested your latest commit on my server. It's working, thanks.

happy-river committed:
mastodon: add ability to build forks without semantic versions

Support Mastodon forks such as glitch-soc which do not
use version numbers (which are required by yarn2nix) by
allowing the maintainer to supply one to update.sh.

happy-river committed:
nixos/mastodon: use glitch-soc as example fork in documentation

matthiasbeyer commented:
> I tested your latest commit on my server. It's working, thanks.

So you mean you have a running mastodon server that you can use? Because if this is true, I am really happy and looking forward hosting my own soon! :tada:

nyanloutre commented:
@matthiasbeyer yes I am hosting one, you can message me on @nyanloutre@social.nyanlout.re once you get your server running

erictapen commented:
I'm also running this PR on https://social.erictapen.name/ and it works fine!

happy-river commented:
> looks good, except we still need to ensure that the node package version follows semver (in case we're passing a git hash to the update script and not semver based tags)

I added another argument to the update script so you can now supply it with both a semver and a git hash, and I changed the documentation to use glitch-soc as an example of how to build a fork, showing the use of both arguments.

nyanloutre commented:
I made a daily systemd timer to prune old medias: https://github.com/nyanloutre/nixpkgs/commit/a8c4d12580506a3bad376c1ae77a45a2bc145386

By default it delete medias from remote server older than 7 days (maybe add an option to configure this ?)

Without it the media folder can grow pretty quickly.

matthiasbeyer commented:
> I made a daily systemd timer to prune old medias: [...]
> By default it delete medias from remote server older than 7 days (maybe add an option to configure this ?)

I would love to see this in this PR (not sure about others though). I would recommend making this configurable. Not only the time period, but other things as well (don't know whether there are flags for `tootctl` maybe to limit the prune to run for 30 mins or something like that...).

nyanloutre commented:
> limit the prune to run for 30 mins
It's possible with TimeoutStartSec in the service unit. By default the timeout is infinity for oneshot services, but it's possible to set a value. Also the unit will be considered failed if the timeout is reached.

matthiasbeyer commented:
> > limit the prune to run for 30 mins
> 
> It's possible with TimeoutStartSec in the service unit. By default the timeout is infinity for oneshot services, but it's possible to set a value. Also the unit will be considered failed if the timeout is reached.

Well, that's suboptimal.

I'm not sure anymore. From what [I see](https://docs.joinmastodon.org/admin/tootctl/) there's a number of commandline calls for `tootctl media` we might want timer jobs for:

* `tootctl media remove --days <configurable> --concurrency <configurable>`
* `tootctl media remove-orphans`
* `tootctl media refresh`

... and possibly a lot more. I'd rather have this documented cleanly, including examples, in the service documentation or nixos.wiki so that one can decide how and what to do when setting up a mastodon service. That might be a better option than providing a number of preconfigured (albeit configurable) timers in the service definition, IMO.

Just my 2ct.

nyanloutre commented:
There is also `tootctl statuses remove` that might be useful. It removes old unreferenced statuses (which mean not from someone you are subscribed or that you replied to)

happy-river commented:
Thank you for the suggestions to add services for Mastodon maintenance tasks. I'm inclined to say that this PR is complex enough already and that another PR to make Mastodon instance maintenance easier could be made after this one is merged.

bqv commented:
Agreed. I'd prefer to just see the base module that we can all agree on merged to reduce even more delay

happy-river commented:
Mastodon v3.1.3 has been released, but something's going wrong on the `yarn2nix` side:

```
yarn install v1.22.4
[1/4] Resolving packages...
warning Lockfile has incorrect entry for "kind-of@^3.0.2". Ignoring it.
warning Lockfile has incorrect entry for "kind-of@^3.2.0". Ignoring it.
warning Lockfile has incorrect entry for "kind-of@^5.0.0". Ignoring it.
warning Lockfile has incorrect entry for "kind-of@^4.0.0". Ignoring it.
warning Lockfile has incorrect entry for "kind-of@^3.0.3". Ignoring it.
error Couldn't find any versions for "kind-of" that matches "^3.0.2" in our cache (possible versions are ""). This is usually caused by a missing entry in the lockfile, running Yarn without the --offline flag may help fix this issue.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Error: Couldn't find any versions for "kind-of" that matches "^5.0.0" in our cache (possible versions are ""). This is usually caused by a missing entry in the lockfile, running Yarn without the --offline flag may help fix this issue.
     at MessageError.ExtendableBuiltin (/nix/store/mciq1ywy9z2p942xwpa0c15avyvyzsbb-yarn-1.22.4/libexec/yarn/lib/cli.js:721:66)
    at new MessageError (/nix/store/mciq1ywy9z2p942xwpa0c15avyvyzsbb-yarn-1.22.4/libexec/yarn/lib/cli.js:750:123)
    at NpmResolver.<anonymous> (/nix/store/mciq1ywy9z2p942xwpa0c15avyvyzsbb-yarn-1.22.4/libexec/yarn/lib/cli.js:50632:15)
    at Generator.next (<anonymous>)
    at step (/nix/store/mciq1ywy9z2p942xwpa0c15avyvyzsbb-yarn-1.22.4/libexec/yarn/lib/cli.js:310:30)
    at /nix/store/mciq1ywy9z2p942xwpa0c15avyvyzsbb-yarn-1.22.4/libexec/yarn/lib/cli.js:321:13

```
I've truncated this, but it continues with an error for each version of "kind-of" for which it gave a warning.

happy-river commented:
Here is the `yarn2nix` issue: [Support package.json field resolutions](https://github.com/moretea/yarn2nix/issues/136)

Mic92 commented:
By the way: This pattern could avoid the expression size explosion: https://github.com/NixOS/nixpkgs/pull/87258/files#diff-97ddd5942a260ac035c022c0c57de234R20

ashkitten commented:
failing to build on nixos-unstable, log: https://gist.github.com/ashkitten/37963dea2cedb6ee22c089a22b2e992a

erictapen commented:
Btw the pattern proposed by @Mic92 seems to not be affected by the `resolutions` field in `package.yaml`, as `code-server` [contains the field](https://github.com/cdr/code-server/blob/3.4.1/package.json).

So maybe we could avoid https://github.com/moretea/yarn2nix/issues/136 and the large expression size at once?

spacekitteh commented:
Any update on this?

Mic92 commented:
Because there seems demand for this module, why maintaining it in nix-community as an out-of-tree NixOS module until a better solution for nodejs is found?

happy-river committed:
mastodon: mastodon: 3.1.2 -> 3.2.0

happy-river commented:
Updated to the latest version of Mastodon, 3.2.0, and rebased on Nixpkgs master.  I solved the yarn2nix resolutions problem by adding a patch which moves the single affected dependency from the resolutions section into the dependencies section of `package.json`. This makes the package build and the two NixOS tests for it pass.  I haven't tried to update my instance with it yet.

Since this is the second Mastodon version in a row in which I've needed to patch something, I added a `--patches` option to `update.sh` to make it easier next time.

Mic92 commented:
I don't want to discourage good work but this PR still adds 0.5 mb nix expressions that everyone needs to download and sometimes evaluate. Even my laptop with 16 GB sometimes fail to do a local evaluation which takes 4 GB at the time - more than smaller devices (rpi4) have. @happy-river I sent you an invite for the nix-community org. We could leave some documentation in the nixpkgs itself for better visibility. Maybe we could also arrange a jobset on hydra so packages end up in the binary cache (if this is necessary).

matthiasbeyer commented:
> 0.5 mb nix expressions

Sounds like we want a dedicated overlay only for mastodon!

happy-river commented:
The problem is not this PR, the problem is that there is no deduplication of the output of `yarn2nix` and `bundix` in Nixpkgs.  Since early February, the last time this topic came up in this discussion, the 6 other packages which use `yarn2nix` (yarn2nix-moretea, riot, gitlab, codimd, prometheus and gotify) have collectively added 533 new automatically generated yarn package definitions to Nixpkgs.  About half are duplicates; my tally of duplicate yarn package definitions currently in Nixpkgs is 3,719.  This PR would add 526 new unique yarn package definitions.  

There are more users of `bundix` and fewer gem definitions, but still about half of the automatically generated ruby gem package definitions already in Nixpkgs are duplicates.

Here's my tool for counting definitions, if you would like to try it yourself: https://github.com/happy-river/dedup

Mic92 commented:
> The problem is not this PR, the problem is that there is no deduplication of the output of `yarn2nix` and `bundix` in Nixpkgs. Since early February, the last time this topic came up in this discussion, the 6 other packages which use `yarn2nix` (yarn2nix-moretea, riot, gitlab, codimd, prometheus and gotify) have collectively added 533 new automatically generated yarn package definitions to Nixpkgs. About half are duplicates; my tally of duplicate yarn package definitions currently in Nixpkgs is 3,719. This PR would add 526 new unique yarn package definitions.
> 
> There are more users of `bundix` and fewer gem definitions, but still about half of the automatically generated ruby gem package definitions already in Nixpkgs are duplicates.
> 
> Here's my tool for counting definitions, if you would like to try it yourself: [happy-river/dedup](https://github.com/happy-river/dedup)

But it's not possible to merge those definition, no? It would be fine we could merge all those definitions some somehow...

happy-river commented:
> But it's not possible to merge those definition, no? It would be fine we could merge all those definitions some somehow...

Let's continue the discussion at #83499.

Church- commented:
@happy-river So using your module I am noticing a number of errors.

Getting this log output currently:
```
Sep 15 17:42:56 teapot.chatsubo.cafe puma[19873]: [d9b08ebd-7fff-4924-a5d4-8df49dea8c59] method=GET path=/packs/js/locale_en-215a622fb8029a05c56b.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=20.21 view=18.34 db=0.00
Sep 15 17:42:56 teapot.chatsubo.cafe puma[19873]: [77fcf7b9-8c5a-4b87-9c27-36dab82fc064] method=GET path=/avatars/original/missing.png format=png controller=ApplicationController action=raise_not_found status=404 duration=3.18 view=1.55 db=0.00
Sep 15 17:42:56 teapot.chatsubo.cafe puma[19873]: [faafb03d-9372-425b-8708-22ad13980dfa] method=GET path=/packs/css/default-4709f3b3.chunk.css format=css controller=ApplicationController action=raise_not_found status=404 duration=38.82 view=37.18 db=0.00
Sep 15 17:43:00 teapot.chatsubo.cafe puma[19873]: [86efa2d1-e315-41c4-a7aa-5eeef2736def] method=GET path=/apple-touch-icon.png format=png controller=ApplicationController action=raise_not_found status=404 duration=56.37 view=4.83 db=0.00
Sep 15 17:43:00 teapot.chatsubo.cafe puma[19873]: [f55acd7f-a719-4f0b-92e4-8d808c310e23] method=GET path=/favicon.ico format=html controller=ApplicationController action=raise_not_found status=404 duration=2.26 view=1.53 db=0.00
Sep 15 17:43:01 teapot.chatsubo.cafe puma[19873]: [b381feeb-5163-48e7-a842-4fd2d92ce0d4] method=GET path=/packs/js/common-df8cc46ac74920496a3a.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=4476.36 view=4468.28
Sep 15 17:43:01 teapot.chatsubo.cafe puma[19873]: [fcf4ad07-978f-4154-8ece-a6001eaec219] method=GET path=/packs/media/images/preview-5df98290371ead9a70bc3cd4733bbfa7.jpg format=jpeg controller=ApplicationController action=raise_not_found status=404 duration=4496.62 view=>
Sep 15 17:43:01 teapot.chatsubo.cafe puma[19873]: [93060c6e-12fb-411d-9875-389811b6393d] method=GET path=/inert.css format=css controller=ApplicationController action=raise_not_found status=404 duration=4651.13 view=4650.16
Sep 15 17:43:01 teapot.chatsubo.cafe puma[19873]: [ffedb120-c16f-46f1-8091-2db220bc943f] method=GET path=/packs/js/locale_en-215a622fb8029a05c56b.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=1.39 view=0.63 db=0.00
Sep 15 17:43:02 teapot.chatsubo.cafe puma[19873]: [c77164f7-c387-4ebb-8e12-0add0e7e6d64] method=GET path=/packs/js/public-8ef3a7c2f6ca686cc121.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=5335.02 view=5333.75
Sep 15 17:43:02 teapot.chatsubo.cafe puma[19873]: [c497efa1-a268-4f57-9e11-fb1ba67060fb] method=GET path=/packs/css/common-9f44d918.css format=css controller=ApplicationController action=raise_not_found status=404 duration=5395.57 view=5394.36
Sep 15 17:43:02 teapot.chatsubo.cafe puma[19873]: [f73be228-97b9-4141-b826-1af7d68c11ca] method=GET path=/packs/media/images/preview-5df98290371ead9a70bc3cd4733bbfa7.jpg format=jpeg controller=ApplicationController action=raise_not_found status=404 duration=3.88 view=1.85
Sep 15 17:43:02 teapot.chatsubo.cafe puma[19873]: [f8dc80d7-79c7-4029-a192-87249376978e] method=GET path=/avatars/original/missing.png format=png controller=ApplicationController action=raise_not_found status=404 duration=3.21 view=1.41
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [6e1613fb-c726-4c10-8d38-98892320efae] method=GET path=/about format=html controller=AboutController action=show status=200 duration=291.94 view=289.41 db=42.79
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [096797a0-cf8f-4895-bd3b-4bbb9c5342bf] method=GET path=/packs/css/common-9f44d918.css format=css controller=ApplicationController action=raise_not_found status=404 duration=3.94 view=2.37 db=0.00
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [63d0467a-8704-4bbc-a629-f0536807bac6] method=GET path=/packs/js/common-df8cc46ac74920496a3a.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=3.30 view=1.47 db=0.00
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [5ba0d500-b4cd-42f0-ae68-d7ea1364f6fa] method=GET path=/packs/css/default-4709f3b3.chunk.css format=css controller=ApplicationController action=raise_not_found status=404 duration=9.32 view=7.92 db=0.00
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [f7d4dbd5-8846-4706-9b1f-1f2bed0fda6b] method=GET path=/packs/js/public-8ef3a7c2f6ca686cc121.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=4.20 view=3.43 db=0.00
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [82e80fc0-332e-47f8-94eb-88eb90bca85b] method=GET path=/packs/js/locale_en-215a622fb8029a05c56b.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=1.88 view=1.11 db=0.00
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [99062724-e544-4be9-85b8-c36010ea41de] method=GET path=/avatars/original/missing.png format=png controller=ApplicationController action=raise_not_found status=404 duration=1.56 view=0.75 db=0.00
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [43473401-3504-42a4-80b2-f804eea7e36f] method=GET path=/inert.css format=css controller=ApplicationController action=raise_not_found status=404 duration=5.06 view=4.30 db=0.00
Sep 15 17:43:12 teapot.chatsubo.cafe puma[19873]: [41df2a37-e5f8-4a39-9ba6-92847bef760c] method=GET path=/packs/media/images/preview-5df98290371ead9a70bc3cd4733bbfa7.jpg format=jpeg controller=ApplicationController action=raise_not_found status=404 duration=2.52 view=1.7>
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [564ff73e-f27f-4fe8-8b53-86d9936f8374] method=GET path=/ format=html controller=HomeController action=index status=302 duration=2.37 view=0.00 db=0.00 location=https://social.chatsubo.cafe/about
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [3a12375d-6f5c-4a24-9a2a-a2a3c4611e09] method=GET path=/about format=html controller=AboutController action=show status=200 duration=97.12 view=91.09 db=4.34
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [8769d090-324d-4ca0-bbbb-e2d52db4ba81] method=GET path=/packs/css/common-9f44d918.css format=css controller=ApplicationController action=raise_not_found status=404 duration=3.28 view=1.14 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [0d645360-7d38-4b4f-8d6e-9445d6c8ca8a] method=GET path=/inert.css format=css controller=ApplicationController action=raise_not_found status=404 duration=2.30 view=1.13 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [bf79a097-0d2f-4bdd-a1ee-cbf4d0fe31c7] method=GET path=/packs/media/images/preview-5df98290371ead9a70bc3cd4733bbfa7.jpg format=jpeg controller=ApplicationController action=raise_not_found status=404 duration=2.22 view=1.1>
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [3548f1ae-6f74-4e5b-81dc-c3f206bfea90] method=GET path=/packs/css/default-4709f3b3.chunk.css format=css controller=ApplicationController action=raise_not_found status=404 duration=7.95 view=5.69 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [e5d63e99-24ea-4788-bf92-a64ac901f844] method=GET path=/avatars/original/missing.png format=png controller=ApplicationController action=raise_not_found status=404 duration=1.72 view=0.84 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [cb0d1595-21ca-4e93-b733-50e609f69787] method=GET path=/packs/js/public-8ef3a7c2f6ca686cc121.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=3.05 view=2.13 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [84bc2989-abba-46eb-8750-530f115333c3] method=GET path=/packs/js/locale_en-215a622fb8029a05c56b.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=2.03 view=0.94 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [6a521591-04a9-4f5a-8091-d301515a36db] method=GET path=/packs/js/common-df8cc46ac74920496a3a.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=2.31 view=1.16 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [72bacd27-19fb-42ab-b045-87845d253d4c] method=GET path=/packs/js/locale_en-215a622fb8029a05c56b.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=4.14 view=2.11 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [37417a1f-6b3c-44a9-b2a5-921b6246dfd4] method=GET path=/inert.css format=css controller=ApplicationController action=raise_not_found status=404 duration=3.14 view=1.23 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [3700958c-3282-42e6-aebb-b9f4ac74606c] method=GET path=/packs/js/public-8ef3a7c2f6ca686cc121.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=2.06 view=0.68 db=0.00
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [894303cb-39d5-4e5c-aba8-a6ee9e632f32] method=GET path=/packs/media/images/preview-5df98290371ead9a70bc3cd4733bbfa7.jpg format=jpeg controller=ApplicationController action=raise_not_found status=404 duration=2.37 view=1.1>
Sep 15 17:43:19 teapot.chatsubo.cafe puma[19873]: [0b166c3e-6a49-4c38-b168-d250a3d92f68] method=GET path=/avatars/original/missing.png format=png controller=ApplicationController action=raise_not_found status=404 duration=2.39 view=1.16 db=0.00
Sep 15 17:43:48 teapot.chatsubo.cafe puma[19873]: [00a18858-70c7-4472-88a0-529409e5fe92] method=POST path=/auth format=html controller=Auth::RegistrationsController action=create status=200 duration=187.36 view=34.78 db=62.78
Sep 15 17:43:48 teapot.chatsubo.cafe puma[19873]: [0c80b899-8bec-4680-a23e-18f475ebde48] method=GET path=/packs/css/common-9f44d918.css format=css controller=ApplicationController action=raise_not_found status=404 duration=3.32 view=1.43 db=0.00
Sep 15 17:43:48 teapot.chatsubo.cafe puma[19873]: [83b4530c-10cc-42bb-89f3-07438b54a288] method=GET path=/packs/css/default-4709f3b3.chunk.css format=css controller=ApplicationController action=raise_not_found status=404 duration=5.09 view=3.29 db=0.00
Sep 15 17:43:48 teapot.chatsubo.cafe puma[19873]: [f0f3f044-b6c7-43fd-b877-ae6aae3123e8] method=GET path=/packs/js/common-df8cc46ac74920496a3a.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=2.27 view=1.54 db=0.00
Sep 15 17:43:48 teapot.chatsubo.cafe puma[19873]: [9139c086-2d9c-411f-affa-f4e40cc3fa7c] method=GET path=/packs/js/locale_en-215a622fb8029a05c56b.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=1.97 view=1.21 db=0.00
Sep 15 17:43:48 teapot.chatsubo.cafe puma[19873]: [5a20d57d-31ba-44bc-b65a-87aff8e81285] method=GET path=/packs/js/public-8ef3a7c2f6ca686cc121.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=2.03 view=1.29 db=0.00
Sep 15 17:43:48 teapot.chatsubo.cafe puma[19873]: [83e28968-4d6c-45a5-9b37-66553a220e1d] method=GET path=/inert.css format=css controller=ApplicationController action=raise_not_found status=404 duration=6.16 view=5.45 db=0.00
Sep 15 17:43:49 teapot.chatsubo.cafe puma[19873]: [a0bab2e7-efad-498d-b962-8ea87fcfd1ed] method=GET path=/inert.css format=css controller=ApplicationController action=raise_not_found status=404 duration=3.81 view=1.85 db=0.00
Sep 15 17:46:44 teapot.chatsubo.cafe puma[19873]: [4ca72719-4760-429b-acf5-6aafcda4ab4c] method=GET path=/packs/media/images/preview-5df98290371ead9a70bc3cd4733bbfa7.jpg format=jpeg controller=ApplicationController action=raise_not_found status=404 duration=3.26 view=0.7>
Sep 15 17:46:44 teapot.chatsubo.cafe puma[19873]: [75f27915-4c46-4cde-aa66-c96ea89c1202] method=GET path=/favicon.ico format=html controller=ApplicationController action=raise_not_found status=404 duration=4.32 view=1.90 db=0.00
Sep 15 18:22:01 teapot.chatsubo.cafe puma[19873]: [2e29c251-9cc9-4ac8-b8a4-ac9e0870d490] method=GET path=/ format=html controller=HomeController action=index status=302 duration=2.53 view=0.00 location=https://social.chatsubo.cafe/about
Sep 15 18:22:01 teapot.chatsubo.cafe puma[19873]: [2561bc17-0894-4e9a-ba49-42ccced0abb1] method=GET path=/about format=html controller=AboutController action=show status=200 duration=220.47 view=218.79 db=59.31
Sep 15 18:22:01 teapot.chatsubo.cafe puma[19873]: [02a00546-9ac7-41f5-8386-089a5a7086a4] method=GET path=/packs/js/common-df8cc46ac74920496a3a.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=6.38 view=3.33
Sep 15 18:22:01 teapot.chatsubo.cafe puma[19873]: [755d8ca1-13b0-4437-b6f0-8797dc1ab0ce] method=GET path=/packs/js/locale_en-215a622fb8029a05c56b.chunk.js format=*/* controller=ApplicationController action=raise_not_found status=404 duration=6.75 view=3.68 db=0.00
Sep 15 18:22:01 teapot.chatsubo.cafe puma[19873]: [f4c1dcad-3534-4e33-aafc-020864c440c8] method=GET path=/packs/css/default-4709f3b3.chunk.css format=css controller=ApplicationController action=raise_not_found status=404 duration=3.84 view=1.96
Sep 15 18:22:01 teapot.chatsubo.cafe puma[19873]: [d95841b1-1bd1-4aef-84f4-0c3a5c95f5f3] method=GET path=/packs/css/common-9f44d918.css format=css controller=ApplicationController action=raise_not_found status=404 duration=4.30 view=2.34 db=0.00
```

Only changes made are:
* mastodon-web listens on 0.0.0.0
* mastodon-streaming listens on 0.0.0.0
* nginx is upstream on another server. It's config for my mastodon virtualHost is 1:1 to the version you would auto-generate.

Curious if you've run into a similar error?

nyanloutre commented:
I'm using this module for a long time now, what is blocking it from getting merged ?

mkg20001 commented:
Currently there's a merge conflict with all-tests.nix, likely since the line got modified in another commit. Should be easy to resolve.

Besides that are there any current blockers for this PR?

erictapen committed:
mastodon: 3.2.0 -> 3.2.2

erictapen committed:
mastodon: 3.2.2 -> 3.3.0

Additionally to running

  ./update.sh --ver v3.3.0 --patches ./resolutions.patch

I did

- Resolve merge conflicts in resolutions.patch
- Specified buildInputs for the ruby gem openssl
- Use ruby_2_7, as it is recommended in the release notes [0]
- build mastodon-assets with RAILS_ENV, as it seems to be required. It
  would probably be nice to set it to production someday, but atm it is
  set to development, as production requires OTP_SECRET to be set.

[0] https://github.com/tootsuite/mastodon/releases/tag/v3.3.0

erictapen commented:
I just upgraded to `v3.3.0`. I tested this on a small instance and it seems to work. This time the upgrade included some manual steps, see the commit message for details.

I hope this is ok for everyone, that I push to this PR? If anybody uses this in production, you probably need to follow the migration instructions in the [release notes](https://github.com/tootsuite/mastodon/releases/tag/v3.3.0).

erictapen reviewed (approved):
This PR and it's predecessor has been waiting for a long time and has seen a fair bit of community involvement. As far as I followed the discussion the only current issue  (besides a trivial merge conflict) is the increase in closure size, due to lack of duplication in `yarn.nix`.

As the metric for the problem is not the size of the Nixpkgs git repository but the size of the resulting tarball, I'd say this is a problem that could still be solved after this PR is merged.

So I'd vouch to get this merged. It is already getting used by people, it is getting maintained. Both of it would be easier if #78810 would be merged. I won't do it myself though, as I only recently aquired the commit bit and don't trust my judgement enough on this controversial PR.

leo60228 commented:
I get this error when building:
```
Building native extensions. This could take a while...
ERROR:  Error installing /nix/store/2ysdwqd7a3ahp7dmwqxgizry3zpxfqx1-bcrypt-3.1.16.gem:
        ERROR: Failed to build gem native extension.

    current directory: /nix/store/asnccd5av7zqnhnxkx08yahylkfakg11-ruby2.7.2-bcrypt-3.1.16/lib/ruby/gems/2.7.0/gems/bcrypt-3.1.16/ext/mri
/nix/store/sxvw9ja9i1ndxslrbkird76m3jsamjsg-ruby-2.7.2/bin/ruby -I /nix/store/sxvw9ja9i1ndxslrbkird76m3jsamjsg-ruby-2.7.2/lib/ruby/2.7.0 -r ./siteconf20210111-8-3yq1qn.rb extconf.rb
creating Makefile

current directory: /nix/store/asnccd5av7zqnhnxkx08yahylkfakg11-ruby2.7.2-bcrypt-3.1.16/lib/ruby/gems/2.7.0/gems/bcrypt-3.1.16/ext/mri
make "DESTDIR=" clean
current directory: /nix/store/asnccd5av7zqnhnxkx08yahylkfakg11-ruby2.7.2-bcrypt-3.1.16/lib/ruby/gems/2.7.0/gems/bcrypt-3.1.16/ext/mri
make "DESTDIR="
make failedNo such file or directory - make "DESTDIR="

Gem files will remain installed in /nix/store/asnccd5av7zqnhnxkx08yahylkfakg11-ruby2.7.2-bcrypt-3.1.16/lib/ruby/gems/2.7.0/gems/bcrypt-3.1.16 for inspection.
Results logged to /nix/store/asnccd5av7zqnhnxkx08yahylkfakg11-ruby2.7.2-bcrypt-3.1.16/lib/ruby/gems/2.7.0/extensions/x86_64-linux/2.7.0/bcrypt-3.1.16/gem_make.out
```
Nothing I've tried is able to reproduce this other than just building the ruby-bcrypt derivation. breakpointHook works, but I can't find any more details on the issue anywhere in the filesystem.

leo60228 commented:
I had a broken `sandbox-paths` in my `nix.conf`, not sure how I didn't catch that until now...

leo60228 commented:
> Make `pkgs.mastodon.override { ... }` work to build forks of Mastodon.

To make this a bit easier, perhaps `update.sh` should create an `override.nix` that's something like this?

```nix
{ mastodon, callPackage }:
mastodon.override {
  version = import ./version.nix;
  srcOverride = callPackage ./source.nix {};
  dependenciesDir = ./.;
}
```

bqv commented:
Merge conflict :)

erictapen commented:
As nobody else joined in my initiative to get this finally merged (either by approving or merging it), I deduce consensus is not strong enough among the commiters. Still there seems to be great interest in having this. So let's have it as an out-of-tree module?

I started working on a Nix Flake [mastodon-nixos](https://github.com/erictapen/mastodon-nixos), that currently somewhat mirrors the functionality of this PR. I intend to further maintain this Flake and add more functionality in the future (e.g. support for non-flake users). Goal is to move the project into the nix-community org, as I guess that'd be the right place for it. Could a member add me to the org so I can push the repo?

Also I will close this PR now. It is currently based on an commit from August 2020. That means that it uses an old version of ffmpeg and imagemagick, whose should have accumulated multiple security vulnerabilities by now. No one should use this branch without a rebase in production. Feel free to reopen if anyone wants to maintain it, but I currently see not much future in it.

bqv commented:
Oh gosh, thank you @erictapen ! A flake is a great idea

Izorkin commented:
How to fix this error?:
```
mastodon-init-dirs-start[91340]: ERROR: Missing RAILS_ENV environment variable, please set it to "production", "development", or "test".
```
and
```
mastodon-init-db.service: Failed to load environment files: No such file or directory
```

erictapen commented:
So as it turns out there are some people from the nix-community IRC willing to merge this.
I rebased on current master. For that I had to open a new PR at #112898.
